### PR TITLE
[#165345969] Allow bindings to be read-only or read-write

### PR DESCRIPTION
**Pull request by richardTowers**
What
---
Allow bindings to be read-only or read-write

In this commit, we allow users to pass a "permissions" parameter when binding an app or creating a service key, so that access credentials can be created with as limited scope as necessary. The valid permissions are "read-only" and "read-write".
    
We refactored how S3 bucket policies are generated and mutated, so that future bucket policy changes are easier to implement.

How to review
---
1. Code review
2. Deploy the [paas-cf PR](https://github.com/alphagov/paas-cf/pull/1865) to your env and play around with it. Check that permissions are limited appropriately
```
cf create-service aws_s3_bucket my_bucket
cf create-service-key my_bucket my_bucket_ro -c '{"permissions": "read-only"}'
cf create-service-key my_bucket my_bucket_rw -c '{"permissions": "read-write"}' # This is the default
```
3. Also review the [bosh release PR](https://github.com/alphagov/paas-s3-broker-boshrelease/pull/3)

Who can review
---
Not @AP-Hunt or @richardTowers

### CODE REVIEW

**Comment by richardTowers** [comment]
This has already been deployed to `andyhunt`. I've tested with:

```
# Creating the bucket
$ cf create-service aws-s3-bucket default my-bucket

# Creating "service keys" (which uses the Bind operation):
$ cf create-service-key my-bucket my-bucket-rw-key -c '{"permissions":"read-write"}'
$ cf create-service-key my-bucket my-bucket-ro-key -c '{"permissions":"read-only"}'

# Accessing the bucket using RW creds
$ cf service-key my-bucket my-bucket-rw-key
$ AWS_ACCESS_KEY_ID=... AWS_SECRET_ACCESS_KEY=... aws s3 cp README.md s3://paas-s3-broker-andyhunt-.../README.md
# Uploads the README file
$ AWS_ACCESS_KEY_ID=... AWS_SECRET_ACCESS_KEY=... aws s3 cp s3://paas-s3-broker-andyhunt-.../README.md -
# Downloads the README file

# Accessing the bucket using RO creds
$ cf service-key my-bucket my-bucket-ro-key
$ AWS_ACCESS_KEY_ID=... AWS_SECRET_ACCESS_KEY=... aws s3 cp README.md s3://paas-s3-broker-andyhunt-.../README.md
# Fails to upload the README file
$ AWS_ACCESS_KEY_ID=... AWS_SECRET_ACCESS_KEY=... aws s3 cp s3://paas-s3-broker-andyhunt-.../README.md -
# Downloads the README file
```

File: s3/policy/statement_builder.go [hunk 1]
```diff
@@ -0,0 +1,56 @@
...
+	} else if permissionName == ReadWritePermissionName {
+		actions = []string{
+			"s3:GetBucketLocation",
+			"s3:ListBucket",
+			"s3:GetObject",
+			"s3:PutObject",
+			"s3:DeleteObject",
+		}
+	} else {
+		log.Panicf("unknown permission name %s", permissionName)
```
Comment (tlwr) [inline s3/policy/statement_builder.go]: I think the function signature for this should change to `(Statement, error)` and this should be error.Formatf or an error type, and then the corresponding failure case handled in the client.

**Comment by tlwr** [comment]
Looks really good and works perfectly from some manual testing.

I have left a comment about using `log.Panicf` which I think is the only controversy

**Comment by tlwr** [comment]
I've pushed a commit which details my changes, feel free to remove/ignore

File: s3/policy/statement_builder.go (hunk 1 above)
Comment (richardTowers) [reply s3/policy/statement_builder.go]: The theory was that we'd validate this in `validPermissionsRequested`, and by the time we get here it's a programming error if it's not an expected value. Would be nice if the type system were strong enough to enforce this, but y'know, go.

Might be better just to bubble the error, as you say.

**Comment by tlwr** [comment]
Richard and I went through and paired on this, I'll bump the relevant PR in https://github.com/alphagov/paas-s3-broker-boshrelease

**Pull request merged**